packages/runtime/container-runtime/src/gc/gcTelemetryTracker.ts

```diff
@@ -0,0 +1,353 @@
+				const event = {
+					eventName: `${state}Object_${nodeUsageProps.usageType}`,
+					pkg: packagePathToTelemetryProperty(nodeUsageProps.packagePath),
+					stack: generateStack(),
+					...eventProps,
+				};
+
+				// Do not log the inactive object x events as error events as they are not the best signal for
+				// detecting something wrong with GC either from the partner or from the runtime itself.
+				if (state === UnreferencedState.Inactive) {
+					this.mc.logger.sendTelemetryEvent(event);
+				} else {
+					this.mc.logger.sendErrorEvent(event);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Log all new references or outbound routes in the current graph that haven't been explicitly notified to GC.
+	 * The principle is that every new reference or outbound route must be notified to GC via the
+	 * addedOutboundReference method. It it hasn't, its a bug and we want to identify these scenarios.
+	 *
+	 * In more simple terms:
+	 * Missing Explicit References = Current References - Previous References - Explicitly Added References;
+	 *
+	 * @param currentGCData - The GC data (reference graph) from the current GC run.
+	 * @param previousGCData - The GC data (reference graph) from the previous GC run.
+	 * @param explicitReferences - New references added explicity between the previous and the current run.
+	 */
+	public logIfMissingExplicitReferences(
+		currentGCData: IGarbageCollectionData,
+		previousGCData: IGarbageCollectionData,
+		explicitReferences: Map<string, string[]>,
+		logger: ITelemetryLogger,
+	) {
+		for (const [nodeId, currentOutboundRoutes] of Object.entries(currentGCData.gcNodes)) {
+			const previousRoutes = previousGCData.gcNodes[nodeId] ?? [];
+			const explicitRoutes = explicitReferences.get(nodeId) ?? [];
+
+			/**
+			 * 1. For routes in the current GC data, routes that were not present in previous GC data and did not have
+			 * explicit references should be added to missing explicit routes list.
+			 * 2. Only include data store and blob routes since GC only works for these two.
+			 * Note: Due to a bug with de-duped blobs, only adding data store routes for now.
+			 * 3. Ignore DDS routes to their parent datastores since those were added implicitly. So, there won't be
+			 * explicit routes to them.
+			 */
+			const missingExplicitRoutes: string[] = [];
+			for (const route of currentOutboundRoutes) {
+				const nodeType = this.getNodeType(route);
+				if (
+					(nodeType === GCNodeType.DataStore || nodeType === GCNodeType.Blob) &&
+					!nodeId.startsWith(route) &&
+					!previousRoutes.includes(route) &&
+					!explicitRoutes.includes(route)
+				) {
+					missingExplicitRoutes.push(route);
+				}
+			}
+
+			if (missingExplicitRoutes.length > 0) {
+				logger.sendErrorEvent({
+					eventName: "gcUnknownOutboundReferences",
+					gcNodeId: nodeId,
+					gcRoutes: JSON.stringify(missingExplicitRoutes),
+				});
+			}
+		}
+	}
+
+	public async logUnreferencedEvents(logger: ITelemetryLogger) {
```

Comment:
`logPendingEvents`?  Anything to clarify/enforce that this is Summarizer-only?

Reply:
Renamed and added comment.